core/src/org/javarosa/core/model/instance/TreeReference.java

```diff
@@ -629,31 +665,31 @@ public int getContext() {
      * Returns the subreference of this reference up to the level specified.
      *
      * For instance, for the reference:
-     *
      * (/data/path/to/node).getSubreference(2) => /data/path/to
      *
      * Used to identify the reference context for a predicate at the same level
      *
-     * Must be an absolute reference, otherwise will throw IllegalArgumentException
-     *
-     * @param i
-     * @return
+     * @param level number of segments to include in the truncated
+     * sub-reference.
+     * @throws IllegalArgumentException if this object isn't an absolute
+     * reference.
+     * @return A clone of this reference object that includes steps up the
+     * specified level.
      */
     public TreeReference getSubReference(int level) {
         if (!this.isAbsolute()) {
             throw new IllegalArgumentException("Cannot subreference a non-absolute ref");
         }
 
-        //Copy construct
-        TreeReference ret = new TreeReference();
-        ret.refLevel = this.refLevel;
-        ret.contextType = this.contextType;
-        ret.instanceName = this.instanceName;
-        ret.data = new Vector<TreeReferenceLevel>();
+        TreeReference subRef = new TreeReference();
+        subRef.refLevel = this.refLevel;
+        subRef.contextType = this.contextType;
+        subRef.instanceName = this.instanceName;
+
         for (int i = 0; i <= level; ++i) {
-            ret.data.addElement(this.data.elementAt(i));
+            subRef.add(this.data.elementAt(i).shallowCopy());
```

Comment:
Delegating through the 'add' method, instead of data.addElement, since it performs potentially crucial internal manipulations (hashCode = -1, size = -1, etc). 
Was there a (performance) reason that this wasn't being done before?